user/configuration-variables.rst

```diff
@@ -0,0 +1,89 @@
+Configuration variables/context in OpenWISP2
+============================================
+
+What configuration variables are, why do we need them?
+------------------------------------------------------
+
+It's often needed for the configuration to differ a little between
+devices, we may need to use device specific values
+(like MAC address, or UUID) in it, in such situation it's not to possible
+to simply use static templates. 
+The configuration variables feature allows us to include in our config
+some default/automatically assigned variables
+and define our own ones in controller admin.
+
+The set of all configuration variables is known as *context*.
+
+Default context
+---------------
+
+Following variables are added automatically
+to the context of each configuration:
+
+- ``id`` - device's UUID
+- ``key`` - device's OpenWISP key
+- ``name`` - device's name in OpenWISP
+- ``mac_address`` - device's MAC address
+
+Referencing variables in configuration
+--------------------------------------
+
+We can include variables in our configuration by simply writing3
+``{{ variable_name }}`` or ``{{variable_name}}`` in it.
+
+Usage example
+~~~~~~~~~~~~~
+
+Using following template (JSON/advanced view notation)
+we can change OpenWISP connection settings on our devices
+without the need to manually set
+``UUID`` and ``key`` parameters in configuration:
+
+.. code-block:: javascript
+
+    {
+      "openwisp": [
+        {
+          "config_name": "controller",
+          "config_value": "http",
+          "url": "http://controller.myopenwisp.org",
+          "interval": "60",
+          "uuid": "{{ id }}",
+          "key": "{{ key }}"
+        }
+      ]
+    }
```

Comment:
add two screenshots, one in which you show that JSON in the advanced mode UI, another one with the resulting configuration visible in the preview.

Reply:
PS: do this as a template, this config is usually added into a template and reused by all devices or group of devices.

Reply:
I mentioned that it's a template (line 37), removing word 'configuration' not to misslead

Reply:
Are you sure we really need screenshots there? They're going to be screenshots of just text.

Reply:
Quoting:

> one in which you show that JSON in the advanced mode UI, another one with the resulting configuration visible in the preview (point out also how to open the preview, it's a button at the top right of the page).

This is not just text. It's showing an example configuration before configuration variables are evaluated and after. Users want to see, they don't want to read. The image will communicate all the written texts in instants. It's very powerful, do not underestimate the power of images in communication. It's the MOST important thing ever.